ChronoKeeper/LogIngestionQueue.h

```diff
@@ -0,0 +1,50 @@
+#ifndef INGESTION_QUEUE_H
+#define INGESTION_QUEUE_H
+
+
+#include <deque>
+
+//INNA: TODO : this is just a draft implementation 
+//
+// IngestionQueue is a funnel into the MemoryDataStore
+// stl deque guarantees O(1) time for addidng elements and resizing 
+// (vector of vectors implementation)
+// I also intend to have only one ingestion thread adding time chunks and elements to this dequeue
+// so no mutex is needed to protect log_event ingestion 
+
+namespace chronolog
+{
+typedef uint64_t StoryId;
+typedef uint64_t ClientId;
+
+struct LogEvent
+{
+	LogEvent(StoryId const& story_id, ClientId const& client_id, uint64_t time, std::string const& record)
+		: storyId(story_id), clientId(client_id)
+		, timestamp(time),logRecord(record)
+	{}  
+	StoryId  storyId;
+	ClientId clientId;
+	uint64_t timestamp;
+	std::string logRecord;
+};
+
+class LogIngestionQueue
+{
+
+public:
+	LogIngestionQueue() {}
+	~LogIngestionQueue() = default;
+
+	void ingestLogEvent( LogEvent const& logEvent)
+	{ eventDeque.push_back(logEvent); }
+
```

Comment:
Do you plan to store events in a queue and do collision check/out-of-order check etc later? If so, do you want to add an API to sort the events according to timestamps? Currently, the code does not have a function to sort or remove duplicates. Another option is to maintain a separate ordered map with just the timestamps. One could use the ordered map to check collisions, perform range searches/interval deletes etc. Either you augment the queue with an ordered map or use a queue+sorting function.

Reply:
I do not plan to deal with timestamp collisions at the ingestion stage .  They would be handled  later in the pipeline as we've discussed at the most recent meeting